airbyte-integrations/connectors/source-shortio/source_shortio/source.py

```diff
@@ -0,0 +1,266 @@
+
+import requests
+import json
+import datetime
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+from airbyte_cdk.logger import AirbyteLogger
+
+class BasicAuthenticator(TokenAuthenticator):
+
+    def get_auth_header(self) -> Mapping[str, Any]:
+        if self.auth_method is None:
+            return {self.auth_header: f"{self._token}"}
+        else:
+            return {self.auth_header: f"{self.auth_method} {self._token}"}
+
+
+class Links(HttpStream, ABC):
+
+    url_base = "https://api.short.io/api"
+    limit = 150
+    primary_key = "id"
+    before_id = None
+    def next_page_token(
+        self, 
+        response: requests.Response
+        ) -> Optional[Mapping[str, Any]]:
+        
+        links = json.loads(response.text)['links']
+        try:
+            id_string = sorted(links, key=lambda k: k['createdAt'], reverse=False)[0]['idString']
+            if self.before_id != id_string:
+                self.before_id = id_string        
+                return id_string
+            else:
+                return None
+        except IndexError:
+            return None
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {
+            "limit" : self.limit,
+            "domain_id" : self._config['domain_id'],
+            "before": next_page_token or None
+        }
+
+    def path(
+        self, 
+        stream_state: Mapping[str, Any] = None, 
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        # Get all the links 
+        return "/links"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        The short.io API can be incosistent in its inclusion of UTM parameters.
+        Here we specifically check if they've been provided and in case they haven't attempt to fetch from the URL directly.
+        """
+        utm_params = {
+            # Passing secondary UTM Campaign in order to capture either or of the 2 args.
+            'utmSource' : 'utm_source',
+            'utmMedium' : 'utm_medium',
+            'utmCampaign' : 'utm_campaign',
+            'utmCampaignId' : 'utm_id',
+            'utmTerm' : 'utm_term', 
+            'utmContent' : 'utm_content'
+        }
+        links = json.loads(response.text)['links']
+        for index, item in enumerate(links):
+            for k,v in utm_params.items():
+                if k not in item.keys():
+                    param = v + '='
+                    original_url = item['originalURL']
+                    # Link may not have all the necessary UTM attributes
```

Comment:
we already note earlier that we do this because the link doeesn't have all the necessary attributes, so we don't need to repeat it here and can remove this comment